Panic in recalculateFirstForSubj

### What version were you using?

[v2.9.22](https://github.com/nats-io/nats-server/releases/tag/v2.9.22)

### What environment was the server running in?

Kubernetes/Linux

### Is this defect reproducible?

I'm not sure how to reproduce this, but this out of bounds access may be related to the fix that was put in for this issue? https://github.com/nats-io/nats-server/issues/4445
https://github.com/nats-io/nats-server/commit/8865c2a703886356769bcbf877546010b79c3523
Maybe this caused the out of bounds access

```
[1] 2023/09/13 15:35:12.278427 [DBG] JETSTREAM - JetStream connection closed: Client Closed
panic: runtime error: slice bounds out of range [-3785495:]

goroutine 30 [running]:
github.com/nats-io/nats-server/v2/server.(*msgBlock).recalculateFirstForSubj(0xc01d4b2680, {0xc023bedb80, 0x77}, 0xba431, 0xc0126d6500)
    github.com/nats-io/nats-server/v2/server/filestore.go:5979 +0x2ca
github.com/nats-io/nats-server/v2/server.(*fileStore).firstSeqForSubj(0xc00aa80580, {0xc023bedb80, 0x77})
    github.com/nats-io/nats-server/v2/server/filestore.go:2584 +0x1c6
github.com/nats-io/nats-server/v2/server.(*fileStore).storeRawMsg(0xc00aa80580, {0xc023bedb80, 0x77}, {0xc02339efc0, 0x205, 0x205}, {0xc01bfef8c0, 0x293, 0x293}, 0xc422b, ...)
    github.com/nats-io/nats-server/v2/server/filestore.go:2389 +0x675
github.com/nats-io/nats-server/v2/server.(*fileStore).StoreMsg(0xc00aa80580, {0xc023bedb80, 0x77}, {0xc02339efc0, 0x205, 0x205}, {0xc01bfef8c0, 0x293, 0x293})
    github.com/nats-io/nats-server/v2/server/filestore.go:2453 +0x155
github.com/nats-io/nats-server/v2/server.(*stream).processJetStreamMsg(0xc00a96ae00, {0xc023bedb80, 0x77}, {0xc27923c480, 0x14}, {0xc02339efc0, 0x205, 0x205}, {0xc01bfef8c0, 0x293, ...}, ...)
    github.com/nats-io/nats-server/v2/server/stream.go:4051 +0x1d1d
github.com/nats-io/nats-server/v2/server.(*stream).internalLoop(0xc00a96ae00)
    github.com/nats-io/nats-server/v2/server/stream.go:4474 +0x1025
created by github.com/nats-io/nats-server/v2/server.(*stream).setupSendCapabilities
    github.com/nats-io/nats-server/v2/server/stream.go:4353 +0x1bc
Stream closed EOF for redacted
```

### Given the capability you are leveraging, describe your expectation?

N/A

### Given the expectation, what is the defect you are observing?

The server panicked

--- a/server/filestore.go
+++ b/server/filestore.go
@@ -3114,9 +3114,7 @@ func (fs *fileStore) firstSeqForSubj(subj string) (uint64, error) {
 			mb.mu.Unlock()
 			return 0, err
 		}
-		ss := mb.fss[subj]
-		mb.mu.Unlock()
-		if ss != nil {
+		if ss := mb.fss[subj]; ss != nil {
 			// Adjust first if it was not where we thought it should be.
 			if i != start {
 				if info, ok := fs.psim[subj]; ok {
@@ -3126,8 +3124,10 @@ func (fs *fileStore) firstSeqForSubj(subj string) (uint64, error) {
 			if ss.firstNeedsUpdate {
 				mb.recalculateFirstForSubj(subj, ss.First, ss)
 			}
+			mb.mu.Unlock()
 			return ss.First, nil
 		}
+		mb.mu.Unlock()
 	}
 	return 0, nil
 }
@@ -6578,6 +6578,7 @@ func (mb *msgBlock) recalculateFirstForSubj(subj string, startSeq uint64, ss *Si
 			return
 		}
 	}
+
 	// Mark first as updated.
 	ss.firstNeedsUpdate = false
 	startSeq++

--- a/server/filestore_test.go
+++ b/server/filestore_test.go
@@ -29,6 +29,7 @@ import (
 	"os"
 	"path/filepath"
 	"reflect"
+	"sync"
 	"testing"
 	"time"
 
@@ -5427,6 +5428,60 @@ func TestFileStoreSyncIntervals(t *testing.T) {
 	checkSyncFlag(false)
 	fs.StoreMsg("Z", nil, []byte("hello"))
 	checkSyncFlag(false)
+}
+
+// https://github.com/nats-io/nats-server/issues/4529
+// Run this wuth --race and you will see the unlocked access that probably caused this.
+func TestFileStoreRecalcFirstSequenceBug(t *testing.T) {
+	fcfg := FileStoreConfig{StoreDir: t.TempDir()}
+	fs, err := newFileStore(fcfg, StreamConfig{Name: "zzz", Subjects: []string{"*"}, MaxMsgsPer: 2, Storage: FileStorage})
+	require_NoError(t, err)
+	defer fs.Stop()
+
+	msg := bytes.Repeat([]byte("A"), 22)
+
+	for _, subj := range []string{"A", "A", "B", "B"} {
+		fs.StoreMsg(subj, nil, msg)
+	}
+	// Make sure the buffer is cleared.
+	clearLMBCache := func() {
+		fs.mu.RLock()
+		mb := fs.lmb
+		fs.mu.RUnlock()
+		mb.mu.Lock()
+		mb.clearCacheAndOffset()
+		mb.mu.Unlock()
+	}
+
+	clearLMBCache()
+
+	// Do first here.
+	fs.StoreMsg("A", nil, msg)
+
+	var wg sync.WaitGroup
+	start := make(chan bool)
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		<-start
+		for i := 0; i < 1_000; i++ {
+			fs.LoadLastMsg("A", nil)
+			clearLMBCache()
+		}
+	}()
+
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		<-start
+		for i := 0; i < 1_000; i++ {
+			fs.StoreMsg("A", nil, msg)
+		}
+	}()
+
+	close(start)
+	wg.Wait()
 }
 
 ///////////////////////////////////////////////////////////////////////////

